File: src/ruby/matmul.rb
# frozen_string_literal: true

def matgen(n)
  tmp = 1.0 / n / n
  Array.new(n) { |i|
    Array.new(n) { |j|
      tmp * (i - j) * (i + j)
    }
  }
end

def matmul(a, b)
  m = a.length
  n = a[0].length
  p = b[0].length
  c = Array.new(m) { Array.new(p, 0.0) }
  for i in 0...m
    ci = c[i]
    ai = a[i]
    k = 0
    while k < n
      aik = ai[k]
      bk = b[k]
      j = 0
      while j < p
        ci[j] += aik * bk[j]
        j += 1
      end
      k += 1
    end
  end
  c
end

n = 1500
if ARGV.length > 0
  n = ARGV[0].to_i
end
n = n / 2 * 2
a = matgen(n)
b = matgen(n)
c = matmul(a, b)
puts c[n / 2][n / 2]

File: src/perl/matmul.pl
#!/usr/bin/perl -w

use strict;
use warnings;

my $n = $ARGV[0] || 1500;
$n = int($n/2) * 2;
my $a = matgen($n);
my $b = matgen($n);
my $x = matmul($a, $b);
print $x->[$n/2][$n/2], "\n";

exit;

sub matgen {
	my ($n) = @_;
	my $a;
	my $tmp = 1 / $n / $n;
	for my $i (0 .. $n - 1) {
		for my $j (0 .. $n - 1) {
			$a->[$i][$j] = $tmp * ($i - $j) * ($i + $j);
		}
	}
	return $a;
}

sub matmul {
	my ($a, $b) = @_;
	my $x;
	my $n = @{$a->[0]};
	my $p = @{$b->[0]};

	for my $i (0 .. @$a - 1) {
		my $xi = $x->[$i] = [ () x $n ];
		for my $k (0 .. $n - 1) {
			my $aik = $a->[$i][$k];
			my $bk  = \@{$b->[$k]};
			for my $j (0 .. $p - 1) {
				$xi->[$j] += $aik * $bk->[$j];
			}
		}
	}
	return $x;
}
